File: apps/server/src/graphql/resolvers/notification.ts
import { and, eq, sql } from 'drizzle-orm';
import { createD1Client } from '../../db';
import * as schema from '../../db/schema';
import type { ContextType } from '../../types';

export const notificationResolvers = {
  Query: {
    notifications: async (
      _: unknown,
      { limit = 20, offset = 0, includeRead = false },
      context: ContextType,
    ) => {
      if (!context.user) {
        throw new Error('Not authenticated');
      }

      const db = createD1Client(context.env);

      const conditions = [eq(schema.notification.userId, context.user.id)];

      if (!includeRead) {
        conditions.push(eq(schema.notification.isRead, 0));
      }

      return await db
        .select()
        .from(schema.notification)
        .where(and(...conditions))
        .orderBy(sql`${schema.notification.createdAt} DESC`)
        .limit(limit)
        .offset(offset);
    },

    unreadNotificationCount: async (_: unknown, __: unknown, context: ContextType) => {
      if (!context.user) {
        throw new Error('Not authenticated');
      }

      const db = createD1Client(context.env);

      const result = await db
        .select({ count: sql`count(*)` })
        .from(schema.notification)
        .where(
          and(eq(schema.notification.userId, context.user.id), eq(schema.notification.isRead, 0)),
        )
        .get();

      return result?.count || 0;
    },
  },

  Mutation: {
    markNotificationRead: async (_: unknown, { id }: { id: string }, context: ContextType) => {
      if (!context.user) {
        throw new Error('Not authenticated');
      }

      const db = createD1Client(context.env);

      await db
        .update(schema.notification)
        .set({ isRead: 1 })
        .where(
          and(eq(schema.notification.id, id), eq(schema.notification.userId, context.user.id)),
        );

      return { success: true };
    },

    markAllNotificationsRead: async (_: unknown, __: unknown, context: ContextType) => {
      if (!context.user) {
        throw new Error('Not authenticated');
      }

      const db = createD1Client(context.env);

      const result = await db
        .update(schema.notification)
        .set({ isRead: 1 })
        .where(
          and(eq(schema.notification.userId, context.user.id), eq(schema.notification.isRead, 0)),
        );

      return {
        success: true,
        count: 0,
      };
    },
  },

  Notification: {
    actionUser: async (parent: { actionUserId?: string }, _: unknown, context: ContextType) => {
      if (!parent.actionUserId) return null;

      const db = createD1Client(context.env);

      return await db
        .select()
        .from(schema.profile)
        .where(eq(schema.profile.userId, parent.actionUserId))
        .get();
    },

    isRead: (parent: { isRead: number }) => Boolean(parent.isRead),
  },
};

File: apps/rust-test-server/src/graphql/resolvers/notification.rs
/*!
# Notification Resolvers

This module handles notification GraphQL operations including:
- Notification retrieval with filtering
- Notification read status management
- Unread notification counts

Based on TypeScript implementation in apps/server/src/graphql/resolvers/notification.ts
*/

use crate::graphql::{
    GraphQLEngine, GraphQLValue, GraphQLError, FieldContext
};
use crate::services::database_simple::DatabaseService;
use std::sync::Arc;
use serde_json::{json, Value};

/// Add notification resolvers to the GraphQL engine
pub fn add_notification_resolvers(engine: &mut GraphQLEngine, _db_service: Arc<DatabaseService>) {
    // TODO: Implement notification resolvers
    // Query resolvers:
    // - notifications(limit, offset, includeRead)
    // - unreadNotificationCount()
    
    // Mutation resolvers:
    // - markNotificationRead(id)
    // - markAllNotificationsRead()
    
    // Type resolvers:
    // - Notification.actionUser
    // - Notification.isRead
}

/// Get notifications for the current user
pub fn get_notifications_resolver(ctx: &FieldContext, _db_service: &DatabaseService) -> Result<GraphQLValue, GraphQLError> {
    let _limit = ctx.field_info.arguments
        .get("limit")
        .and_then(|v| v.as_i64())
        .unwrap_or(20);

    let _offset = ctx.field_info.arguments
        .get("offset")
        .and_then(|v| v.as_i64())
        .unwrap_or(0);

    let include_read = ctx.field_info.arguments
        .get("includeRead")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);

    // TODO: Implement actual notification retrieval with filtering
    let mock_notifications = if include_read {
        vec![
            create_mock_notification("notif_1", "NEW_FOLLOW", false),
            create_mock_notification("notif_2", "NEW_COMMENT", true),
            create_mock_notification("notif_3", "POST_SAVE", false),
        ]
    } else {
        vec![
            create_mock_notification("notif_1", "NEW_FOLLOW", false),
            create_mock_notification("notif_3", "POST_SAVE", false),
        ]
    };

    Ok(GraphQLValue::List(mock_notifications.into_iter().map(GraphQLValue::from).collect()))
}

/// Get count of unread notifications
pub fn get_unread_notification_count_resolver(_ctx: &FieldContext, _db_service: &DatabaseService) -> Result<GraphQLValue, GraphQLError> {
    // TODO: Implement actual unread count
    Ok(GraphQLValue::Int(2))
}

/// Mark a specific notification as read
pub fn mark_notification_read_resolver(ctx: &FieldContext, _db_service: &DatabaseService) -> Result<GraphQLValue, GraphQLError> {
    let notification_id = ctx.field_info.arguments
        .get("id")
        .and_then(|v| v.as_str())
        .ok_or_else(|| GraphQLError::new("Missing required argument 'id'"))?;

    // TODO: Implement actual notification read marking
    // - Update notification isRead status
    // - Return updated count
    
    let response = json!({
        "success": true,
        "count": 1
    });

    Ok(GraphQLValue::from(response))
}

/// Mark all notifications as read for current user
pub fn mark_all_notifications_read_resolver(_ctx: &FieldContext, _db_service: &DatabaseService) -> Result<GraphQLValue, GraphQLError> {
    // TODO: Implement actual bulk notification read marking
    // - Update all unread notifications for current user
    // - Return total count updated
    
    let response = json!({
        "success": true,
        "count": 5
    });

    Ok(GraphQLValue::from(response))
}

/// Get action user for notification (type resolver)
pub fn get_notification_action_user(_notification_id: &str, _action_user_id: Option<&str>, _db_service: &DatabaseService) -> Result<Option<GraphQLValue>, GraphQLError> {
    // TODO: Implement actual action user retrieval
    if let Some(_user_id) = _action_user_id {
        let user = json!({
            "id": "profile_1",
            "userId": "user_1",
            "username": "actionuser",
            "profileImage": null,
            "bio": "Action user bio",
            "verifiedType": "none",
            "isPrivate": false,
            "isFollowing": false,
            "followersCount": 10,
            "followingCount": 5,
            "createdAt": "2024-01-01T00:00:00Z",
            "updatedAt": "2024-01-01T00:00:00Z",
            "isBlocked": false
        });
        Ok(Some(GraphQLValue::from(user)))
    } else {
        Ok(None)
    }
}

// Helper functions

/// Create a mock notification for testing
fn create_mock_notification(id: &str, notification_type: &str, is_read: bool) -> Value {
    let (message, resource_type) = match notification_type {
        "NEW_FOLLOW" => ("started following you", "user"),
        "NEW_COMMENT" => ("commented on your post", "post"),
        "COMMENT_REPLY" => ("replied to your comment", "comment"),
        "MENTION" => ("mentioned you in a post", "post"),
        "POST_SAVE" => ("saved your post", "post"),
        "FOLLOW_REQUEST" => ("requested to follow you", "user"),
        _ => ("sent you a notification", "unknown"),
    };

    json!({
        "id": id,
        "type": notification_type,
        "actionUser": {
            "id": "profile_action",
            "username": "actionuser",
            "profileImage": null
        },
        "resourceId": "resource_123",
        "resourceType": resource_type,
        "message": message,
        "isRead": is_read,
        "createdAt": "2024-01-01T00:00:00Z"
    })
}

/// Generate notification message based on type and context
pub fn generate_notification_message(notification_type: &str, action_username: &str, resource_type: Option<&str>) -> String {
    match notification_type {
        "NEW_FOLLOW" => format!("{} started following you", action_username),
        "NEW_COMMENT" => format!("{} commented on your post", action_username),
        "COMMENT_REPLY" => format!("{} replied to your comment", action_username),
        "MENTION" => format!("{} mentioned you in a post", action_username),
        "POST_SAVE" => format!("{} saved your post", action_username),
        "FOLLOW_REQUEST" => format!("{} requested to follow you", action_username),
        _ => format!("{} sent you a notification", action_username),
    }
}

/// Create notification for various events
pub async fn create_notification(
    _user_id: &str,
    _notification_type: &str,
    _action_user_id: Option<&str>,
    _resource_id: Option<&str>,
    _resource_type: Option<&str>,
    _db_service: &DatabaseService
) -> Result<(), GraphQLError> {
    // TODO: Implement actual notification creation
    // - Generate notification ID
    // - Create notification record
    // - Send real-time updates if needed
    Ok(())
}